programs/utilities/bots.sc

```diff
@@ -0,0 +1,293 @@
+
+  modify(p, 'flying', data:'f');
+  modify(p, 'gamemode', data:'g');
+
+  if (data:'m', (
+    mounted = filter(in_dimension(data:'d', entity_area('*', data:'m', [0.5, 0.5, 0.5])), _~'type' != 'player');
+    if (mounted, (
+      schedule(1, _(p, m) -> modify(p, 'mount', m), p, mounted:0);
+    ), (
+      schedule(5, _(p, m) -> modify(p, 'pos', m), p, data:'m');
+      schedule(6, _(n) -> run(str('player %s mount', n)), data:'n');
+    ));
+  ));
+
+  run(str('player %s %s', data:'n', if(data:'c', 'sneak', 'unsneak')));
+
+  selected_slot = data:'s'||0;
+  modify(p, 'selected_slot', selected_slot);
+  _set_hand(p, selected_slot, data:'r');
+  _set_hand(p, -1, data:'l');
+
+  if (data:'at', run(str('player %s attack continuous', data:'n')));
+  if (data:'u', run(str('player %s use continuous', data:'n')));
+);
+
+apply_set(set) -> (
+  if (!_check_permission(), _error('Not allowed'));
+
+  for (global_bots:set, (
+    p = player(_:'n');
+
+    if (p, (
+      if (p~'player_type' != 'fake', continue());
+      run(str('/player %s stop', _:'n'));
+      run(str('/execute in %s run teleport %s %f %f %f %f %f', _:'d', _:'n', _:'x', _:'y', _:'z', _:'a', _:'p'));
+    ),(
+      run(str('player %s spawn at %f %f %f facing %f %f in %s', _:'n', _:'x', _:'y', _:'z', _:'a', _:'p', _:'d'));
+    ));
+
+    __configure_bot_when_available(_);
+  ));
+
+  _print_formatted(_set_name(set), 'w  applied.');
+);
+
+global_to_delete = null;
+
+delete_set(set) -> (
+  if (!_check_permission(), _error('Not allowed'));
+
+  if (set == null, (
+	if (global_to_delete != null, _print_formatted(_set_name(global_to_delete), 'y  not deleted.'));
+    global_to_delete = null;
+
+  ), global_to_delete == set, (
+    global_to_delete = null;
+    delete(global_bots, set);
+    _persist();
+    _print_formatted(_set_name(set), 'r  deleted.');
+
+  ), !has(global_bots, set), (
+    global_to_delete = null;
+    _print_formatted(_set_name(set), 'r  does not exist.');
+
+  ), (
+    global_to_delete = set;
+
+    _print_formatted(
+      'w Are you sure you want to delete ', _set_name(set), 'w ? ',
+      'lb [YES] ', str('!/%s delete %s', global_app_name, set),
+      'rb [NO]', str('!/%s delete', global_app_name),
+    );
+  ));
+);
+
+kill_set(set) -> (
+  if (!_check_permission(), _error('Not allowed'));
+
+  for (global_bots:set, kill(_:'n'));
+  _print_formatted('w Bots in ', _set_name(set), 'w  killed.');
+);
+
+kill_all() -> (
+  if (!_check_permission(), _error('Not allowed'));
+
+  for (filter(player('all'), _~'player_type' == 'fake'), kill(_~'name'));
+  _print_formatted(_bot_name('All'), 'w  bots killed.');
+);
+
+kill(bot) -> (
+  run('player ' + bot + ' stop');
+  run('player ' + bot + ' dismount');
+  run('player ' + bot + ' kill');
+);
+
+_on_exit() -> (
+  save_set('#autosave');
+  for (filter(player('all'), _~'player_type' == 'fake'), (
+    run('player ' + _~'name' + ' dismount');
+  ));
+);
+
+if (run('whitelist list'), ( //multiplayer
```

Comment:
This is kind of hacky but works. The `whitelist` command only exists on servers so `run` returns `null` in singleplayer worlds.
Is there a better way to do this?

Reply:
Maybe some options of `system_info` under the server category can be used in a similar way, where they only return `null` when not in multiplayer. `server_whitelist` also exists there.
https://github.com/gnembon/fabric-carpet/blob/master/docs/scarpet/Full.md#server-related-properties

Reply:
Thanks for the tip! I now use `system_info('server_ip') != null`.
(Still a bit hacky but much better)